bip-signmessage-redone.mediawiki

```diff
@@ -0,0 +1,156 @@
+
+== Background ==
+
+* Assume two actors, a prover <code>P</code> and a verifier <code>V</code>.
+* <code>P</code> wants to prove that they own the private key <code>k</code> associated with a given address <code>A</code> (which in turn is derived from the pubkey <code>kG</code>).
+* Let <code>V</code> generate a message <code>M</code> and hand this to <code>P</code>.
+* <code>P</code> generates a signature <code>S</code> by signing the message <code>M</code> using <code>k</code>. Given <code>S</code>, <code>V</code> can prove that <code>P</code> has the private key associated with <code>A</code>.
+
+The astute reader will notice that the above is missing a critical part, namely the pubkey <code>kG</code>, without which the verifier cannot actually verify the message. The current message signing standard solves this via a cryptographic trick, wherein the signature <code>S</code> above is a special "recoverable signature" type. Given the message <code>M</code> and the signature <code>S</code>, it is then possible to recover the pubkey <code>kG</code>. The system thus derives the address for the pubkey <code>kG</code>, and if it does not match <code>A</code>, the proof is deemed invalid.
+
+While this is a neat trick, it unnecessarily restricts and complicates the message signing mechanism; for instance, it is currently not possible to sign a message for a P2SH address, because there is no pubkey to recover from the resulting signature.
+
+== Motivation ==
+
+The current message signing standard only works for P2PKH (1...) addresses. By extending it to use a Bitcoin Script based approach, it could be made more generic without causing a too big burden on implementers, who most likely have access to Bitcoin Script interpreters already.
+
+== Specification ==
+
+This BIP follows the specification of BIP-325 challenges and solutions.
+
+Let there be two virtual transactions to_spend and to_sign.
+
+The "to_spend" transaction is:
+
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = 0000...000
+    vin[0].prevout.n = 0xFFFFFFFF
+    vin[0].nSequence = 0
+    vin[0].scriptSig = OP_0 PUSH32[ message_hash ]
+    vin[0].scriptWitness = []
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = message_challenge
+
+where message_hash is a BIP340-tagged hash of the message, i.e. sha256_tag(m), where tag = "BIP????-signed-message", and message_challenge is the to be proven (public) key script.
+
+The "to_sign" transaction is:
+
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = to_spend.txid
+    vin[0].prevout.n = 0
+    vin[0].nSequence = 0
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = message_signature
+
+It may include other inputs, to facilitate a proof of funds.
+
+A message signature is considered valid, inconclusive, or invalid based on whether the to_sign transaction is a valid spend of the to_spend transaction or not, according to the following steps (also see Consensus and standard flags section further down):
```

Comment:
If the signature check is performed by the script interpreter, it will honor sighash flags. I think the document should specify that the signature must include SIGHASH_ALL flag, otherwise signature with SIGHASH_NONE would verify with any `message_challenge`

Reply:
Also, `scriptSig` is not included in segwit signature hash, rather the script is included. It may not be clear for the implementors how they should make the signature commit to `message_hash`. I think the spec should say that the signature should be with sighash calculated with:
```
script = OP_0 PUSH32[ message_hash ]
hash_type = SIGHASH_ALL
amount = 0
sigversion = SIGVERSION_WITNESS_V0
```

(witness sigversion will allow to avoid stripping `CODESEPARATOR` that happens with sighash calculating with `sigversion = SIGVERSION_BASE`)

Reply:
Thanks, yep. I think requiring it to be SIGHASH_ALL is beneficial, since this also supports proof of funds.

Reply:
For some reason I didn't see your 2nd response until now. I added a note that they must use `SIGHASH_ALL`, which commits to the message hash as expected. What do you think?

Reply:
Moved the conversation about the rest to the `vin[0].scriptSig = OP_0 PUSH32[ message_hash ]` line.